### DISCUSSION TITLE
Event handler support for receiving events in batches

### DISCUSSION THREAD

**Discussion by slashdotdash** [commanded/commanded#575] [Ideas]
Event handlers receive events one at a time for processing. Returning `:ok` indicates successful handling, the event is acknowledged, then the next event can be received.

```elixir
defmodule ExampleHandler do
  use Commanded.Event.Handler, name: __MODULE__

  def handle(event, _metadata) do
    # Process single event
    :ok
  end
end
```

## Proposal

Add support to event handlers allowing configuration of the batch size of events to receive.  This would be the maximum number of events that would be processed if there is a backlog of events. If the handler is up to date then events will be processed immediately, it won't wait for an entire batch size worth of events.

```elixir
defmodule ExampleHandler do
  use Commanded.Event.Handler, name: __MODULE__, batch_size: 50
end
```

### How to handle a batch of events?

1. Use the existing `handle/2` function, passing either a single event (batch size 1) or a list of events (batch size > 1):

    ```elixir
    defmodule ExampleHandler do
      use Commanded.Event.Handler, name: __MODULE__, batch_size: 1
      
      def handle(event, _metadata) do
        # Process single event (batch size = 1)      
        :ok
      end
    end
    ```

    ```elixir
    defmodule ExampleHandler do
      use Commanded.Event.Handler, name: __MODULE__, batch_size: 50
      
      def handle(events, _metadata) do
        # Process list of events (batch size > 1)      
        :ok
      end
    end
    ```

2. Extend the event handler behaviour to add a `handle_batch/2` event:

    ```elixir
    defmodule ExampleHandler do
      use Commanded.Event.Handler, name: __MODULE__, batch_size: 1
           
      def handle(event, _metadata) do
        # Process single event (batch size = 1)
        :ok
      end
    end
    ```

    ```elixir
    defmodule ExampleHandler do
      use Commanded.Event.Handler, name: __MODULE__, batch_size: 50
      
      def handle_batch(events, _metadata) do
        # Process list of events (batch size > 1)      
        :ok
      end
    end
    ```

    This would be similar to how Broadway handles batches of messages in its [`handle_batch/4`](https://hexdocs.pm/broadway/Broadway.html#c:handle_batch/4) callback function.

**Comment by justgage** [comment]
I really want/need this. I think the main thing that would be nice would be able to return _how many_ you were able to process successfully? Kind of like returning a new offset. Like `{:ok, 12}`? That's probably not the best API but something like that.

**Comment by larskluge** [comment]
I'm keen about this feature too.. :)

**Comment by davydog187** [comment]
I will be taking this on :)

**Comment by davydog187** [comment]
@slashdotdash If I'm not mistaken, I remember you saying that there was already built-in EventStore support for reading in batches. From what I can tell, an `Event.Handler` already receives [events in batches](https://github.com/commanded/commanded/blob/master/lib/commanded/event/handler.ex#L722), and then reduces over the events with `handle_event/2`.

In order to tackle this ticket, could you please help shed some light on the following questions:

1. `Event.Handler`s rely on a [persistent subscription](https://github.com/commanded/eventstore/blob/bc5d6dd6e4f1d54efc667b66197486fe79f8f087/guides/Subscriptions.md#persistent-subscriptions) to the event store. From what I can tell, only `stream_forward` and `stream_all_forward` support configurable batch sizes. Do I need to implement that first in `EventStore`? Am I missing something here?
2. We need to decide between choice 1 and 2 for the API. I am partial to choice 2, adding a `handle_batch` function, and if a user configures the event handler at compile time with a `batch_size`, we should raise if they do not implement `handle_batch/2`


What do you think?

**Comment by cdegroot** [comment]
FWIW, partial to having a separate handle_batch. I think Kafka makes you return the last successfully processed offset, and that could work here as well. +1 to the idea of batching, it took us three months to go through `$all` the last time we tried with regular event handlers :)

**Comment by mbnissen** [comment]
I would love to have this feature

**Comment by cdegroot** [comment]
Heads up: [we](https://metrist.io) decided we want this enough to implement it, so that's what we are going to work on in the coming time.

**Comment by davydog187** [comment]
@cdegroot take a look at https://github.com/commanded/commanded/pull/471 which might be helpful for implementing, happy to answer questions as well

**Comment by cdegroot** [comment]
Just a heads up that we started work on this, shouldn't be too long before we turn this into a PR. Started off @davydog187's branch and a first test runs. Until PR time, you can tag along on https://github.com/Metrist-Software/commanded/tree/batching-support

**Comment by Nezteb** [comment]
Any updates by chance @cdegroot? 😄👀 

![screenshot_2023-11-07_at_17 51 21@2x](https://github.com/commanded/commanded/assets/3588798/76bc1781-e392-42a1-b26c-947e14549439)
